core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -0,0 +1,263 @@
+        shouldCallTerminated = true;
+      }
+    }
+    listener.transportShutdown(
+        Status.OK.withDescription("Channel requested transport to shut down"));
+    if (shouldCallTerminated) {
+      listener.transportTerminated();
+    }
+  }
+
+  /**
+   * Shuts down this transport and cancels all streams that it owns, hence immediately terminates
+   * this transport.
+   */
+  public void shutdownNow(Status status) {
+    shutdown();
+    Collection<PendingStream> savedPendingStreams = null;
+    synchronized (lock) {
+      if (pendingStreams != null) {
+        savedPendingStreams = pendingStreams;
+        pendingStreams = null;
+      }
+    }
+    if (savedPendingStreams != null) {
+      for (PendingStream stream : savedPendingStreams) {
+        stream.setError(status);
+      }
+      listener.transportTerminated();
+    }
+    // If savedPendingStreams == null, transportTerminated() has already been called in shutdown().
+  }
+
+  /**
+   * Transfers all the pending and future requests and pings to the given transport.
+   *
+   * <p>May only be called after {@link #start(Listener)}.
+   *
+   * <p>{@code transport} will be used for all future calls to {@link #newStream}, even if this
+   * transport is {@link #shutdown}.
+   */
+  public void setTransport(ClientTransport transport) {
+    boolean savedShutdown;
+    synchronized (lock) {
+      Preconditions.checkState(delegate == null, "setTransport already called");
+      Preconditions.checkState(listener != null, "start() not called");
+      delegate = Preconditions.checkNotNull(transport, "transport");
+      savedShutdown = shutdown;
+      for (PendingPing ping : pendingPings) {
+        ping.createRealPing(transport);
+      }
+      pendingPings = null;
+      if (pendingStreams != null) {
+        for (PendingStream stream : pendingStreams) {
+          stream.createRealStream(transport);
+        }
+      }
+      pendingStreams = null;
+    }
+    if (savedShutdown) {
+      listener.transportTerminated();
+    } else {
+      listener.transportReady();
+    }
+  }
+
+  @VisibleForTesting
+  int getPendingStreamsCount() {
+    synchronized (lock) {
+      return pendingStreams == null ? 0 : pendingStreams.size();
+    }
+  }
+
+  private class PendingStream extends DelayedStream {
+    private final MethodDescriptor<?, ?> method;
+    private final Metadata headers;
+
+    private PendingStream(MethodDescriptor<?, ?> method, Metadata headers) {
+      this.method = method;
+      this.headers = headers;
+    }
+
+    private void createRealStream(ClientTransport transport) {
+      setStream(transport.newStream(method, headers));
+    }
+
+    // TODO(zhangkun83): DelayedStream.setError() doesn't have a clearly-defined semantic to be
+    // overriden. Make it clear or find another method to override.
+    @Override
+    void setError(Status reason) {
+      synchronized (lock) {
+        if (pendingStreams != null) {
+          pendingStreams.remove(this);
```

Comment:
If `shutdown == true` and `pendingStreams.isEmpty()` then `pendingStreams = null` and notify listener about termination.